MDANSE/Src/MDANSE/Framework/QVectors/DispersionLatticeQVectors.py

```diff
@@ -40,32 +45,39 @@ class DispersionLatticeQVectors(LatticeQVectors):
     )
 
     def _generate(self):
-        start = self._configuration["start"]["value"]
-        direction = self._configuration["direction"]["value"]
+        start = self._configuration["start"]["value"].array
+        step_vector = self._configuration["step"]["value"].array
         n_steps = self._configuration["n_steps"]["value"]
 
         hkls = np.array(start)[:, np.newaxis] + np.outer(
-            direction, np.arange(0, n_steps)
+            step_vector, np.arange(0, n_steps)
         )
-
-        # The k matrix (3,n_hkls)
         vects = self.hkl_to_qvectors(hkls, self._unit_cell)
+        hkl_vectors, weights = self.lattice_vectors_with_weights(vects, self._unit_cell)
+        # The k matrix (3,n_hkls)
+        q_vectors = self.hkl_to_qvectors(hkl_vectors, self._unit_cell)
 
-        dists = np.sqrt(np.sum(vects**2, axis=0))
+        dists = np.linalg.norm(q_vectors, axis=0)
+        sign_array = np.sign(np.dot(step_vector, hkl_vectors))
+        keyvals = dists.copy()
+        keyvals[np.where(sign_array < 0)] *= -1
```

Comment:
Would this not be simpler (plus no where [O(N)], intermediate array and contiguous memory) as:

```suggestion
        keyvals *= sign_array
```

Reply:
While this would be simpler, it would also replace the any vector orthogonal to the step direction with a 0.
If I start from (2,0,-10) along (0,0,1), the code without the `where` will generate a sequence with ..., (2,0,-1), (0,0,0), (2,0,1), ...

Reply:
What about
```python
        keyvals[sign_array < 0] *= -1
```